File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/cli/verbs/deps.mjs
import { resolve } from "node:path";
import { buildRegistry } from "../../registry.mjs";
import { topoSort, findMissingTargets } from "../../deps/toposort.mjs";
import { checkInversions } from "../../deps/inversion.mjs";
import { checkConflicts, checkUnpinned } from "../../deps/conflict.mjs";

function parseArgs(args) {
  const opts = { path: null, format: "human", strict: false, extraTiers: [] };
  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a === "--format") opts.format = args[++i];
    else if (a === "--strict") opts.strict = true;
    else if (a === "--extra-tiers")
      opts.extraTiers = (args[++i] ?? "")
        .split(",")
        .map((s) => s.trim())
        .filter(Boolean);
    else if (!a.startsWith("--") && opts.path == null) opts.path = a;
  }
  return opts;
}

export async function deps(args) {
  const opts = parseArgs(args);
  if (!opts.path) {
    return {
      exitCode: 64,
      output: "usage: skills-ref deps <root> [--format json|human] [--strict] [--extra-tiers ...]\n",
    };
  }
  const reg = buildRegistry(resolve(opts.path));
  const { order, cycles } = topoSort(reg);
  const missing = findMissingTargets(reg);
  const inversions = checkInversions(reg);
  const conflicts = checkConflicts(reg);
  const unpinned = checkUnpinned(reg);

  let exitCode = 0;
  if (cycles.length > 0 || inversions.length > 0) exitCode = 1;
  else if (missing.length > 0) exitCode = 2;
  else if (conflicts.length > 0) exitCode = 1;
  else if (opts.strict && unpinned.length > 0) exitCode = 2;

  if (opts.format === "json") {
    const body = sortKeys({
      order,
      cycles,
      missing,
      inversions,
      conflicts,
      unpinned,
      warnings: reg.warnings,
    });
    return { exitCode, output: JSON.stringify(body) + "\n" };
  }
  let out = "";
  out += `order: ${order.join(" -> ") || "(empty)"}\n`;
  if (cycles.length) {
    out += "cycles:\n";
    for (const c of cycles) out += `  - ${c.join(" -> ")}\n`;
  }
  if (missing.length) {
    out += "missing in-repo targets:\n";
    for (const m of missing) out += `  - ${m.from} -> ${m.to}\n`;
  }
  if (inversions.length) {
    out += "inversions:\n";
    for (const i of inversions)
      out += `  - ${i.from} (${i.fromTier}) -> ${i.to} (${i.toTier}) [downward dependency forbidden]\n`;
  }
  if (conflicts.length) {
    out += "conflicts:\n";
    for (const c of conflicts) out += `  - ${c.message}\n`;
  }
  if (unpinned.length) {
    out += "unpinned (WARN):\n";
    for (const u of unpinned) out += `  - ${u.message}\n`;
  }
  if (reg.warnings.length) {
    out += "registry warnings:\n";
    for (const w of reg.warnings) out += `  - ${w}\n`;
  }
  out += `\nsummary: ${cycles.length} cycle(s), ${inversions.length} inversion(s), ${missing.length} missing, ${conflicts.length} conflict(s), ${unpinned.length} unpinned\n`;
  return { exitCode, output: out };
}

function sortKeys(v) {
  if (Array.isArray(v)) return v.map(sortKeys);
  if (v && typeof v === "object") {
    const o = {};
    for (const k of Object.keys(v).sort()) o[k] = sortKeys(v[k]);
    return o;
  }
  return v;
}

File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/verbs/deps.py
from __future__ import annotations

import json
from pathlib import Path

from ..deps.conflict import check_conflicts, check_unpinned
from ..deps.inversion import check_inversions
from ..deps.toposort import find_missing_targets, topo_sort
from ..registry import build_registry


def deps(args) -> tuple[int, str]:
    root = Path(args.path).resolve()
    reg = build_registry(str(root))
    ts = topo_sort(reg)
    order = ts["order"]
    cycles = ts["cycles"]
    missing = find_missing_targets(reg)
    inversions = check_inversions(reg)
    conflicts = check_conflicts(reg)
    unpinned = check_unpinned(reg)

    exit_code = 0
    if cycles or inversions:
        exit_code = 1
    elif missing:
        exit_code = 2
    elif conflicts:
        exit_code = 1
    elif getattr(args, "strict", False) and unpinned:
        exit_code = 2

    if args.format == "json":
        body = {
            "order": order,
            "cycles": cycles,
            "missing": missing,
            "inversions": inversions,
            "conflicts": conflicts,
            "unpinned": unpinned,
            "warnings": list(reg.warnings),
        }
        return exit_code, json.dumps(_sort(body), sort_keys=True, separators=(",", ":")) + "\n"

    parts = []
    parts.append(f"order: {' -> '.join(order) if order else '(empty)'}\n")
    if cycles:
        parts.append("cycles:\n")
        for c in cycles:
            parts.append(f"  - {' -> '.join(c)}\n")
    if missing:
        parts.append("missing in-repo targets:\n")
        for m in missing:
            parts.append(f"  - {m['from']} -> {m['to']}\n")
    if inversions:
        parts.append("inversions:\n")
        for i in inversions:
            parts.append(
                f"  - {i['from']} ({i['fromTier']}) -> {i['to']} ({i['toTier']}) "
                f"[downward dependency forbidden]\n"
            )
    if conflicts:
        parts.append("conflicts:\n")
        for c in conflicts:
            parts.append(f"  - {c['message']}\n")
    if unpinned:
        parts.append("unpinned (WARN):\n")
        for u in unpinned:
            parts.append(f"  - {u['message']}\n")
    if reg.warnings:
        parts.append("registry warnings:\n")
        for w in reg.warnings:
            parts.append(f"  - {w}\n")
    parts.append(
        f"\nsummary: {len(cycles)} cycle(s), {len(inversions)} inversion(s), "
        f"{len(missing)} missing, {len(conflicts)} conflict(s), {len(unpinned)} unpinned\n"
    )
    return exit_code, "".join(parts)


def _sort(v):
    if isinstance(v, list):
        return [_sort(x) for x in v]
    if isinstance(v, dict):
        return {k: _sort(v[k]) for k in sorted(v.keys())}
    return v
